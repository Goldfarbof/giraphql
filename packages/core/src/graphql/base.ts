import { GraphQLType } from 'graphql';
import BasePlugin from '../plugin';
import BuildCache from '../build-cache';

export default abstract class BaseType<
  Types extends GiraphQLSchemaTypes.TypeInfo,
  Name extends string | keyof Types['Input'] | keyof Types['Output'],
  Shape,
  InputShape = Shape,
  MatchShape = Shape
> {
  typename: Name;

  shape?: Shape;

  inputShape?: InputShape;

  matchShape?: MatchShape;

  abstract kind: 'Object' | 'Union' | 'Interface' | 'Enum' | 'Scalar' | 'InputObject';

  constructor(name: Name, shape?: Shape) {
    this.typename = name;
    this.shape = shape;
  }

  abstract buildType(cache: BuildCache<Types>, plugins: BasePlugin<Types>[]): GraphQLType;
}
