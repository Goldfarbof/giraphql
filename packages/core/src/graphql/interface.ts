import { GraphQLInterfaceType } from 'graphql';
import fromEntries from 'object.fromentries';
import BaseType from './base';
import { ShapeFromTypeParam, TypeParam, InterfaceName, ObjectName } from '../types';
import Field from '../field';
import FieldBuilder from '../fieldUtils/builder';
import ObjectType from './object';
import BasePlugin from '../plugin';
import BuildCache from '../build-cache';

export default class InterfaceType<
  Shape extends {},
  Types extends GiraphQLSchemaTypes.TypeInfo,
  Name extends InterfaceName<Types>
> extends BaseType<Types, Name, ShapeFromTypeParam<Types, Name, true>> {
  kind: 'Interface' = 'Interface';

  description?: string;

  options: GiraphQLSchemaTypes.InterfaceTypeOptions<{}, Types, any>;

  fieldShape?: Shape;

  constructor(name: Name, options: GiraphQLSchemaTypes.InterfaceTypeOptions<Shape, Types, Name>) {
    super(name);

    this.description = options.description;

    this.options = options;
  }

  getFields() {
    return this.options.shape(new FieldBuilder({}, this.typename));
  }

  buildType(cache: BuildCache<Types>, plugins: BasePlugin<Types>[]): GraphQLInterfaceType {
    let types: ObjectType<{}, any, Types, ObjectName<Types>>[];

    return new GraphQLInterfaceType({
      name: String(this.typename),
      description: this.description,
      resolveType: (obj: unknown) => {
        if (!types) {
          types = cache.getImplementers(this.typename);
        }

        for (const type of types) {
          if (type.isType(obj)) {
            return String(type.typename);
          }
        }

        return String(this.typename);
      },
      fields: () =>
        fromEntries(
          Object.entries(cache.getFields(this.typename)).map(([key, field]) => [
            key,
            (field as Field<{}, Types, TypeParam<Types>, TypeParam<Types>>).build(
              key,
              cache,
              plugins,
            ),
          ]),
        ),
    });
  }
}
