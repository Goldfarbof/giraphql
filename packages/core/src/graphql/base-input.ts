import BaseType from './base';

export default abstract class BaseInputType<Shape = unknown, InputShape = Shape> extends BaseType<
  Shape
> {
  shape?: Shape;

  inputShape?: InputShape;

  abstract kind: 'Enum' | 'Scalar' | 'InputObject';
}
