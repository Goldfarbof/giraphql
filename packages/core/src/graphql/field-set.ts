import { FieldsShape } from '../types';

export default class FieldSet<Types extends GiraphQLSchemaTypes.TypeInfo, Shape = unknown> {
  kind: 'FieldSet' = 'FieldSet';

  forType: string;

  shape: FieldsShape<any, any>;

  constructor(type: string, shape: FieldsShape<Types, Shape>) {
    this.forType = type;
    this.shape = shape;
  }
}
