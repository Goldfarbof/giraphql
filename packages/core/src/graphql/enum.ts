import { GraphQLEnumType, GraphQLEnumValueConfigMap, GraphQLEnumValueConfig } from 'graphql';
// @ts-ignore
import fromEntries from 'object.fromentries';
import BaseType from './base';
import { EnumValues, ShapeFromEnumValues } from '../types';

export default class EnumType<
  Types extends GiraphQLSchemaTypes.TypeInfo,
  Name extends string,
  Values extends EnumValues = EnumValues,
  Shape extends string = ShapeFromEnumValues<Values>
> extends BaseType<Types, Name, Shape> {
  values: GraphQLEnumValueConfigMap;

  description?: string;

  kind: 'Enum' = 'Enum';

  options: GiraphQLSchemaTypes.EnumTypeOptions<EnumValues>;

  constructor(name: Name, options: GiraphQLSchemaTypes.EnumTypeOptions<Values>) {
    super(name);

    this.options = options;

    this.values = Array.isArray(options.values)
      ? fromEntries(options.values.map(key => [key, {}]))
      : fromEntries(
          Object.entries(options.values)
            .map(([key, value]) => {
              if (value && typeof value === 'object') {
                return [key, value];
              }

              if (typeof value === 'string') {
                return [value, {}];
              }

              return null;
            })
            .filter(Boolean) as [string, GraphQLEnumValueConfig][],
        );

    this.description = options.description;
  }

  buildType() {
    return new GraphQLEnumType({
      description: this.description,
      name: this.typename,
      values: this.values,
    });
  }
}
