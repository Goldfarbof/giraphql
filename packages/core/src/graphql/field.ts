/* eslint-disable @typescript-eslint/consistent-type-assertions */
import { GraphQLFieldConfig, GraphQLFieldConfigArgumentMap } from 'graphql';
// @ts-ignore
import fromEntries from 'object.fromentries';
import {
  TypeParam,
  InputFields,
  ShapeFromTypeParam,
  FieldNullability,
  FieldOptionsFromKind,
} from '../types';
import { typeFromParam, buildArg } from '../utils';
import BaseType from './base';
import BasePlugin from '../plugin';
import { BuildCache } from '..';

export default class Field<
  Args extends InputFields<Types>,
  Types extends GiraphQLSchemaTypes.TypeInfo,
  ParentType extends TypeParam<Types>,
  Type extends TypeParam<Types>,
  Nullable extends FieldNullability<Types, Type> = FieldNullability<Types, Type>,
  Extends extends string | null = null,
  Kind extends 'Object' | 'Interface' | 'Root' | 'Subscription' =
    | 'Object'
    | 'Interface'
    | 'Root'
    | 'Subscription',
  Options extends FieldOptionsFromKind<
    Types,
    ParentType,
    Type,
    Nullable,
    Args,
    Kind
  > = FieldOptionsFromKind<Types, ParentType, Type, Nullable, Args, Kind>
> {
  shape?: ShapeFromTypeParam<Types, Type, true>;

  nullable: Nullable;

  args: Args = {} as Args;

  extendsField: Extends;

  type: Type;

  options: Options;

  parentTypename: string;

  constructor(
    options: Options & {
      extendsField?: Extends;
    },
    parentTypename: string,
  ) {
    this.options = options;
    this.nullable = options.nullable || (false as Nullable);
    this.args = options.args ? options.args! : ({} as Args);
    this.extendsField = options.extendsField || (null as Extends);
    this.type = options.type;
    this.parentTypename = parentTypename;
  }

  protected buildArgs(cache: BuildCache<Types>): GraphQLFieldConfigArgumentMap {
    return fromEntries(
      Object.keys(this.args).map(key => {
        const arg = this.args[key];

        return [
          key,
          {
            description:
              typeof arg !== 'object' || arg instanceof BaseType || Array.isArray(arg)
                ? undefined
                : arg.description,
            required:
              typeof arg !== 'object' || arg instanceof BaseType || Array.isArray(arg)
                ? false
                : arg.required || false,
            type: buildArg(arg, cache),
            defaultValue:
              typeof arg !== 'object' || arg instanceof BaseType || Array.isArray(arg)
                ? undefined
                : arg.default,
          },
        ];
      }),
    );
  }

  build(
    name: string,
    cache: BuildCache<Types>,
    plugins: BasePlugin<Types>[],
  ): GraphQLFieldConfig<unknown, unknown> {
    const baseConfig: GraphQLFieldConfig<unknown, unknown> = {
      args: this.buildArgs(cache),
      description: this.options.description,
      resolve:
        cache.resolverMock(this.parentTypename, name) ||
        (this.options as { resolve?: (...args: unknown[]) => unknown }).resolve ||
        (() => {
          throw new Error(`Not implemented: No resolver found for ${this.parentTypename}.${name}`);
        }),
      subscribe:
        cache.subscribeMock(this.parentTypename, name) ||
        (this.options as { subscribe?: (...args: unknown[]) => unknown }).subscribe,
      type: typeFromParam(this.type, cache, this.nullable),
      extensions: this.options.extensions,
    };

    return plugins.reduce((config, plugin) => {
      return plugin.updateFieldConfig
        ? plugin.updateFieldConfig(name, this as any, config, cache)
        : config;
    }, baseConfig);
  }
}
