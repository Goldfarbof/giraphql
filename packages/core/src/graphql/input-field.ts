import { GraphQLInputFieldConfig, GraphQLArgumentConfig } from 'graphql';
// @ts-ignore
import { inputTypeFromParam } from '../utils';
import { BuildCache, InputType, SchemaTypes, partialInputShapeKey, PartialInput } from '..';

export default class InputField<T> implements PartialInput {
  kind: 'InputField' = 'InputField';

  [partialInputShapeKey]: T;

  required: boolean | { list: boolean; items: boolean };

  type: InputType<SchemaTypes> | [InputType<SchemaTypes>];

  options: GiraphQLSchemaTypes.InputOptions;

  constructor(options: GiraphQLSchemaTypes.InputOptions) {
    this.options = options;
    this.required = options.required ?? false;
    this.type = options.type;
  }

  // This is currently used to for fields on Input Objects, and Arguments on Fields
  build(cache: BuildCache): GraphQLInputFieldConfig & GraphQLArgumentConfig {
    return {
      description: this.options.description,
      defaultValue: this.options.defaultValue,
      extensions: this.options.extensions,
      type: inputTypeFromParam(this.type, cache, this.required),
    };
  }
}
