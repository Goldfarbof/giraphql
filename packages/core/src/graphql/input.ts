import { GraphQLInputObjectType, GraphQLInputFieldConfigMap } from 'graphql';
// @ts-ignore
import fromEntries from 'object.fromentries';
import { InputFields } from '../types';
import BaseType from './base';
import { buildArg } from '../utils';
import InputFieldBuilder from '../fieldUtils/input';
import BuildCache from '../build-cache';

export default class InputObjectType<
  Types extends GiraphQLSchemaTypes.TypeInfo,
  Shape,
  Fields extends InputFields<Types>,
  Name extends string,
  MatchShape = Shape
> extends BaseType<Types, Name, Shape, Shape, MatchShape> {
  kind: 'InputObject' = 'InputObject';

  fields: Fields;

  options: GiraphQLSchemaTypes.InputTypeOptions<Types, InputFields<Types>>;

  constructor(name: Name, options: GiraphQLSchemaTypes.InputTypeOptions<Types, Fields>) {
    super(name);

    this.fields = options.shape(new InputFieldBuilder());

    this.options = options;
  }

  buildFields(cache: BuildCache<Types>): GraphQLInputFieldConfigMap {
    return fromEntries(
      Object.keys(this.fields).map(key => {
        const field = this.fields[key];
        return [
          key,
          {
            description:
              typeof field !== 'object' || field instanceof BaseType || Array.isArray(field)
                ? undefined
                : field.description,
            required:
              typeof field !== 'object' || field instanceof BaseType || Array.isArray(field)
                ? false
                : field.required || false,
            type: buildArg(field, cache),
          },
        ];
      }),
    );
  }

  buildType(cache: BuildCache<Types>) {
    return new GraphQLInputObjectType({
      name: this.typename,
      description: this.options.description,
      fields: () => this.buildFields(cache),
      extensions: this.options.extensions,
    });
  }
}
