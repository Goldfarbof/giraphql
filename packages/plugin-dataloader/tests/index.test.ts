import { gql } from 'apollo-server';
import { execute, printSchema } from 'graphql';
import { createContext } from './example/context';
import schema from './example/schema';

describe('dataloader', () => {
  it('generates expected schema', () => {
    expect(printSchema(schema)).toMatchSnapshot();
  });

  describe('queries', () => {
    it('valid queries', async () => {
      const query = gql`
        query {
          counts {
            name
            calls
            loaded
          }
          users {
            id
          }
          user {
            id
          }
          userNodes {
            id
          }
          userNode {
            id
          }
          userNodes2: userNodes {
            id
          }
          userNode2: userNode {
            id
          }
          nodes(ids: ["VXNlck5vZGU6MTIz", "VXNlck5vZGU6NDU2"]) {
            id
          }
          user2: user(id: "2") {
            id
          }
          posts(ids: [123, 456]) {
            id
            title
            content
          }
          posts2: posts(ids: [123, 789]) {
            id
            title
            content
          }
          post(id: 1) {
            id
          }
          post2: post(id: 2) {
            id
          }
          fromContext1 {
            id
          }
          fromContext2 {
            id
          }
          fromContext3 {
            id
          }
          fromContext4 {
            id
          }
        }
      `;

      const result = await execute({
        schema,
        document: query,
        contextValue: createContext(),
      });

      expect(result).toMatchInlineSnapshot(`
        Object {
          "data": Object {
            "counts": Array [
              Object {
                "calls": 1,
                "loaded": 5,
                "name": "users",
              },
              Object {
                "calls": 1,
                "loaded": 3,
                "name": "userNodes",
              },
              Object {
                "calls": 1,
                "loaded": 3,
                "name": "posts",
              },
              Object {
                "calls": 1,
                "loaded": 2,
                "name": "post",
              },
            ],
            "fromContext1": Object {
              "id": "123",
            },
            "fromContext2": Object {
              "id": "456",
            },
            "fromContext3": Object {
              "id": "789",
            },
            "fromContext4": Array [
              Object {
                "id": "123",
              },
              Object {
                "id": "456",
              },
            ],
            "nodes": Array [
              Object {
                "id": "VXNlck5vZGU6MTIz",
              },
              Object {
                "id": "VXNlck5vZGU6NDU2",
              },
            ],
            "post": Object {
              "id": "1",
            },
            "post2": Object {
              "id": "2",
            },
            "posts": Array [
              Object {
                "content": "123 title",
                "id": "123",
                "title": "123 title",
              },
              Object {
                "content": "456 title",
                "id": "456",
                "title": "456 title",
              },
            ],
            "posts2": Array [
              Object {
                "content": "123 title",
                "id": "123",
                "title": "123 title",
              },
              Object {
                "content": "789 title",
                "id": "789",
                "title": "789 title",
              },
            ],
            "user": Object {
              "id": "1",
            },
            "user2": Object {
              "id": "2",
            },
            "userNode": Object {
              "id": "1",
            },
            "userNode2": Object {
              "id": "1",
            },
            "userNodes": Array [
              Object {
                "id": "VXNlck5vZGU6MTIz",
              },
              Object {
                "id": "VXNlck5vZGU6NDU2",
              },
              Object {
                "id": "VXNlck5vZGU6Nzg5",
              },
            ],
            "userNodes2": Array [
              Object {
                "id": "VXNlck5vZGU6MTIz",
              },
              Object {
                "id": "VXNlck5vZGU6NDU2",
              },
              Object {
                "id": "VXNlck5vZGU6Nzg5",
              },
            ],
            "users": Array [
              Object {
                "id": "123",
              },
              Object {
                "id": "456",
              },
              Object {
                "id": "789",
              },
            ],
          },
        }
      `);
    });

    it('query with errors', async () => {
      const query = gql`
        query {
          counts {
            name
            calls
            loaded
          }
          users(ids: ["-123", "-456", "789"]) {
            id
          }
          user(id: "-123") {
            id
          }
          userNodes(ids: ["-123", "-456", "789"]) {
            id
          }
          userNode(id: "-123") {
            id
          }
          user2: user(id: "2") {
            id
          }
          posts(ids: [-123, -456, 780]) {
            id
            title
            content
          }
          post(id: -1) {
            id
          }
          post2: post(id: 2) {
            id
          }
        }
      `;

      const result = await execute({
        schema,
        document: query,
        contextValue: createContext(),
      });

      expect(result.data).toMatchInlineSnapshot(`
        Object {
          "counts": Array [
            Object {
              "calls": 1,
              "loaded": 4,
              "name": "users",
            },
            Object {
              "calls": 1,
              "loaded": 3,
              "name": "userNodes",
            },
            Object {
              "calls": 1,
              "loaded": 3,
              "name": "posts",
            },
            Object {
              "calls": 1,
              "loaded": 2,
              "name": "post",
            },
          ],
          "post": null,
          "post2": Object {
            "id": "2",
          },
          "posts": Array [
            null,
            null,
            Object {
              "content": "780 title",
              "id": "780",
              "title": "780 title",
            },
          ],
          "user": null,
          "user2": Object {
            "id": "2",
          },
          "userNode": null,
          "userNodes": Array [
            null,
            null,
            Object {
              "id": "VXNlck5vZGU6Nzg5",
            },
          ],
          "users": Array [
            null,
            null,
            Object {
              "id": "789",
            },
          ],
        }
      `);
      expect(result.errors).toMatchInlineSnapshot(`
        Array [
          [GraphQLError: Invalid ID -123],
          [GraphQLError: Invalid ID -456],
          [GraphQLError: Invalid ID -123],
          [GraphQLError: Invalid ID -456],
          [GraphQLError: Invalid ID -1],
          [GraphQLError: Invalid ID -123],
          [GraphQLError: Invalid ID -123],
          [GraphQLError: Invalid ID -123],
          [GraphQLError: Invalid ID -456],
        ]
      `);
    });
  });
});
