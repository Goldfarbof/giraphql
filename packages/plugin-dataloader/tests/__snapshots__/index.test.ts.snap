// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`dataloader generates expected schema 1`] = `
"type CallCount {
  calls: Int!
  loaded: Int!
  name: String!
}

interface Node {
  id: ID!
}

type Post {
  content: String!
  id: ID!
  title: String!
}

type Query {
  addOnUser: User!
  counts: [CallCount!]!
  fromContext1: User!
  fromContext2: User!
  fromContext3: User!
  fromContext4: [User!]!
  fromContext5: [User!]!
  node(id: ID!): Node
  nodes(ids: [ID!]!): [Node]!
  post(id: Int!): Post
  posts(ids: [Int!]!): [Post]
  user(id: String): User
  userNode(id: String): User
  userNodes(ids: [String!]): [UserNode]
  users(ids: [String!]): [User]
}

interface TestInterface {
  idFromInterface: ID!
}

type User implements TestInterface {
  id: ID!
  idFromInterface: ID!
  self: User!
}

type UserNode implements Node {
  id: ID!
}
"
`;
